keps/sig-storage/3476-volume-group/README.md

```diff
@@ -0,0 +1,1677 @@
+	......
+}
+```
+
+#### VolumeSnapshot and VolumeSnapshotContent
+
+For VolumeSnapshot, we cannot request a VolumeSnapshot to be added to be VolumeGroupSnapshot, therefore VolumeGroupSnapshotName is only in the Status but not in the Spec.
+
+```
+type VolumeSnapshotStatus struct{
+	......
+        // +optional
+        VolumeGroupSnapshotName *string
+	......
+}
+
+type VolumeSnapshotContentStatus struct{
+	......
+        // +optional
+        VolumeGroupSnapshotContentName *string
+	......
+}
+```
+
+### Example Yaml Files
+
+#### Create Volume Group
+
+Example yaml files to create a VolumeGroupClass and a VolumeGroup are in the following.
+
+Create a VolumeGroupClass that supports volumeGroupSnapshot:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupClass
+metadata:
+  name: volumeGroupClass1
+spec:
+  parameters:
+     …...
+ supportVolumeGroupSnapshot: true
+```
+
+Create a VolumeGroup belongs to this VolumeGroupClass:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroup
+metadata:
+  Name: volumeGroup1
+spec:
+  volumeGroupClassName: volumeGroupClass1
+```
+
+#### Add PVC to VolumeGroup
+
+Create a PVC that belongs to the volume group which supports volumeGroupSnapshot:
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: pvc1
+spec:
+  accessModes:
+  - ReadWriteOnce
+  dataSource: null
+  resources:
+	requests:
+  	storage: 1Gi
+  storageClassName: storageClass1
+  volumeMode: Filesystem
+  volumeGroupNames: [volumeGroup1]
+```
+
+#### Create VolumeGroupSnapshot
+
+Create a VolumeGroupSnapshotClass:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshotClass
+metadata:
+  name: volumeGroupSnapshotClass1
+spec:
+  parameters:
+     …...
+```
+
+A VolumeGroupSnapshot taken from the VolumeGroup dynamically:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshot
+metadata:
+  name: my-group-snapshot
+spec:
+  source:
+      volumeGroupName: volumeGroup1
+  volumeGroupSnapshotClassName: volumeGroupSnapshotClass1
+```
+
+A new external VolumeGroup controller will handle VolumeGroupClass, VolumeGroup, and VolumeGroupContent resources. We may need to split this into two controllers, one common controller that handles common functions such as binding, and one sidecar controller that calls the CSI driver.
+
+External provisioner will be modified to read information from volume groups (through volumeGroupNames) and pass them down to the CSI driver.
+
+A new external VolumeGroupSnapshot controller will handle VolumeGroupSnapshotClass, VolumeGroupSnapshot, and VolumeGroupSnapshotContent resources. We may need to split this into two controllers, one common controller that handles common functions such as binding, and one sidecar controller that calls the CSI driver.
+
+Snapshot controller will be modified to update VolumeSnapshot status. External snapshotter sidecar will be modified to update VolumeSnapshotContent status.
+
+### CSI Changes
```

Comment:
In case others have missed it, the PR for the CSI Spec changes is at container-storage-interface/spec#519.

My main concern for the added procedures, is that (almost?) all of them would also require the `secrets` map in order to access the storage system. This is a requirement for Ceph-CSI, and I imagine for other true stateless CSI drivers too.

Reply:
Sure, I can add secrets.

Reply:
Secrets are already in CSI spec. We can handle them the same way that we handle secrets for VolumeSnapshot APIs.